src/Utilities/Compiler/Extensions/IOperationExtensions.cs

```diff
@@ -801,6 +801,348 @@ public static bool IsSetMethodInvocation(this IPropertyReferenceOperation operat
+        {
+            deconstructionAssignment = null;
+
+            var previousOperation = operation;
+            operation = operation.Parent;
+
+            while (operation != null)
+            {
+                switch (operation.Kind)
+                {
+                    case OperationKind.DeconstructionAssignment:
+                        deconstructionAssignment = (IDeconstructionAssignmentOperation)operation;
+                        return deconstructionAssignment.Target == previousOperation;
+
+                    case OperationKind.Tuple:
+                    case OperationKind.Conversion:
+                    case OperationKind.Parenthesized:
+                        previousOperation = operation;
+                        operation = operation.Parent;
+                        continue;
+
+                    default:
+                        return false;
+                }
+            }
+
+            return false;
+        }
+
+        public static RefKind GetRefKind(this IReturnOperation operation, ISymbol containingSymbol)
+        {
+            var containingMethod = TryGetContainingAnonymousFunctionOrLocalFunction(operation) ?? (containingSymbol as IMethodSymbol);
+            return containingMethod?.RefKind ?? RefKind.None;
+        }
+
+        public static IMethodSymbol? TryGetContainingAnonymousFunctionOrLocalFunction(this IOperation? operation)
+        {
+            operation = operation?.Parent;
+            while (operation != null)
+            {
+                switch (operation.Kind)
+                {
+                    case OperationKind.AnonymousFunction:
+                        return ((IAnonymousFunctionOperation)operation).Symbol;
+
+                    case OperationKind.LocalFunction:
+                        return ((ILocalFunctionOperation)operation).Symbol;
+                }
+
+                operation = operation.Parent;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Retursn true if the given operation is a regular compound assignment,
+        /// i.e. <see cref="ICompoundAssignmentOperation"/> such as <code>a += b</code>,
+        /// or a special null coalescing compoud assignment, i.e. <see cref="ICoalesceAssignmentOperation"/>
+        /// such as <code>a ??= b</code>.
+        /// </summary>
+        public static bool IsAnyCompoundAssignment(this IOperation operation)
+        {
+            switch (operation)
+            {
+                case ICompoundAssignmentOperation _:
+                case ICoalesceAssignmentOperation _:
+                    return true;
+
+                default:
+                    return false;
+            }
+        }
+#endif
+    }
+
+    [Flags]
+#pragma warning disable CA1714 // Flags enums should have plural names
+    internal enum ValueUsageInfo
```

Comment:
Nit: consider moving `ValueUsageInfo` and `ValueUsageInfoExtensions` to a separate file.